Tilemap rework

henkhooft opened the pull request:
<!-- The notes within these arrows are for you but can be deleted. -->

## Summary
This PR adds a new system for managing tiles in a tilemap. Numerous large changes are made under the hood that should solve a few fundamental problems that exist with the current implementation.

### Disclaimer
> This is still in the testing fase. I wanted to test the system thoroughly first to remove the majority of the bugs before we get rid of the old tilemap system and port the dependent code.

## Main features and fixes:
**Grid-based design with chunks**
The new design leverages collections more, which allows us to go faster through objects that are on the same layer. Next to that, the map is divided into chunks of 16 x 16 tiles. This should allow more flexibility in the future in cases where we only need to look at objects nearby, without the need to go through the entire map.

**Saving and loading**
The tilemap is now fully serialized and can be saved and loaded at will. This allows a few options like the ability to save chaotic stations at the end of a round for other game modes, but will also make resolving merge conflicts in the map easier. 

**Multi-tile objects**
Objects that are bigger than a single tile can now be successfully added to the map. You can now place a cryo tube again without having to worry about it colliding with other nearby objects.

**Networking is no longer broken**
Maybe one of the biggest improvements compared with the current implementation is that all fixtures that require networking (i.e. jukebox) are succesfully working again. This is due to the difference in how tile objects are networked, with now every object having a NetId. No need to add this to every object though, the tilemap will handle that for you. Also tiles shouldn't disappear for clients anymore...

**Multiple tilemaps**
You are no longer limited to working on a single tilemap. Which means that we can have shuttles now. A new map can be added from the Editor.

## Pictures/Videos
![afbeelding](https://user-images.githubusercontent.com/5231626/124963443-1ae93b00-e020-11eb-91d9-eb7f797988ac.png)
Finally the jukebox is functional again!

<!-- Include photos or videos if possible to help reviewers. -->

<!-- It may also be used in our monthly devblog. -->

## How to use the new tilemap system
1. Add a new scene or open an existing one
2. Add an empty gameobject and add a TileManager script. Make sure you don't use the old one.
3. Open up the new tilemap editor via RE:SS3D Editor tools -> TileMap Editor

## Adding new objects
1. Right click and add a new TileObjectSO. Make sure they exist under a "Resources" folder if you want it to work in the build.

![afbeelding](https://user-images.githubusercontent.com/5231626/124964700-9c8d9880-e021-11eb-8eb8-3b10596fd21e.png)

2. Fill in the name and prefab and select on which layer the object belongs. If the object spans multiple tiles, use the width and height settings.

![afbeelding](https://user-images.githubusercontent.com/5231626/124964850-c5ae2900-e021-11eb-95a4-935e48f44876.png)

3. Remove the existing FixtureStateMaintainer script. Rotations will not work otherwise.

![afbeelding](https://user-images.githubusercontent.com/5231626/124965285-3d7c5380-e022-11eb-9a6d-81952e24036b.png)


## The Editor
![afbeelding](https://user-images.githubusercontent.com/5231626/124964400-43256980-e021-11eb-8fbd-d3d4272b695d.png)
The yellow lines depict the edge of a chunk. Can be toggled on/off under **Grid settings**

## TODO

- [x] Implement simple adjacency connector
- [x] Implement advanced adjacency connector
- [x] Implement offset adjacency connector
- [x] Optional blocked connections
- [x] Optional blocked connections networking
- [x] Editor place UI
- [x] Sub layers (pipes, wall mounts)
- [x] GameObject for each layer
- [x] Add tile/fixture restrictions
- [x] Test networking and fixture sync
- [x] ~~Allow editor to run in play mode~~ Would be a new feature and game mode in the construction system
- [x] Save/Load based on scene name
- [x] Re-implement chunking
- [x] Override existing object in editor
- [x] Test multi-tile objects
- [x] Migrate existing fixtures
- [x] Delete existing tilemap and port dependent code
- [x] Visual indicator for when objects cannot be build
- [x] Rework adjacency networking
- [x] Implement adjacency cross-connect
- [x] Clean redundant/stale code
- [x] Comment code and update final design documentation
- [x] Remove old FixtureStateMaintainer script
- [x] Change color of chunks
- [x] Implement Tile drag handler
- [x] Add connectors to all tables
- [x] Build a simple ministation
- [x] Create multiple Overlay layers
- [x] More plenum restrictions for lattices and catwalks
- [x] Add Specific type for adjacencies
- [x] Remake MiniStation after naming update
- [ ] Add option for custom tileobject serialization

## Bugs

- [x] Icons don't load when switching layer
- [x] Prefabs don't rotate in editor
- [x] Multi tile objects have improper position placement
- [x] Tilemap isn't correctly loaded at start on the build version
- [x] ~~Join mid-round doesn't work. Likely not related to tilemap but to the scene loader~~ Investigated and is due to the order in which scene loading messages are send. Not related to the tilemap
- [x] Editor loads icons very slowly
- [x] Editor doesn't load properly after Unity re-initializes
- [x] Airlocks are not connecting to walls
- [x] Airlock wallcaps are facing the wrong direction
- [x] TileEditor forces focus when placing tiles
- [x] No tile restrictions when deleting objects
- [x] No tile restrictions when overriding objects
- [x] Weird behaviour when adding a second TileManager
- [x] Created objects doesn't spawn on client
- [x] Chunk boundaries don't disappear
- [x] No tile restrictions for walls in place
- [x] Adjacencies not updated on spawned objects
- [x] Adjacencies not updated on clients at start
- [x] Not all meshes are correctly linked to prefabs as their location moved
- [x] X6 and X5 meshes not working properly in the advanced connector
- [x] Different tilemaps can overlap
- [x] Empty map list displays weirdly in the editor
- [x] Icons are not always loaded at start
- [x] Walls and windows lack adjacency connector
- [x] Fix light bulb fixture prefab link
- [x] Disposal Pipe prefab is missing its simple adjacency connector
- [x] Atmos pipes are missing their mesh filters and assigned adjacency meshes
- [x] L2 Atmos pipe has the Pipe Object script but the others don't
- [x] Add the word "Mount" after the "HighWall" and "LowWall" to avoid any confusion
- [x] Walls want to connect to doors from all directions
- [x] Girders won't place on the map
- [x] Capitalize tile objects
- [x] Metal wall prefab and tile object in the /Walls/ folder is duplicate
- [x] Visual indicator not working for allow overwrite
- [x] Rotation of ghost object is not changed when selecting a new one in the editor
- [x] Z-fighting happening when placing overlays
- [x] Overlays with rotations are not found when soft loading
- [x] Missing meshes on walls and plenums
- [x] Walls prevent plenums from connecting correctly
<!-- List ANYTHING not working correctly, either part of your new change, or another part of the game. -->


### Limitations / considerations
- All atmospherics related code is disabled and needs to be reworked.
- ~~Not all icons always load at the start. Reopening the editor solves this.~~
- Prefabs that need to be spawned (i.e. crowbar/wrench) need to be added to the spawnable prefab list manually. This should be resolved when addressable assets are in.
- The wrench and crowbar should be reworked. Currently their interactions are rather limited and lack the ability to set a rotation for example.
- You can see inside of walls through windows. This is an issue with the adjacency connectors for windows.
- Wires do not connect as they have a complex design and need a new type of adjacency connector.
- Items are not part of the tilemap. If wanted, will need to be integrated at a later stage to also allow saving/loading.
<!-- Any known bugs will likely require sorting out before the PR is merged. -->

### Design
![afbeelding](https://user-images.githubusercontent.com/5231626/124963534-3ce2bd80-e020-11eb-95f2-e50f700d1683.png)

More elaborate design document will be made if the design and implementation is successful.

henkhooft committed:
Add initial structure for new tilemap

henkhooft committed:
Add basis for tilemap editor

henkhooft committed:
Add support for multiple tilemaps

henkhooft committed:
Improve editor and saving/loading

henkhooft committed:
Enable placement of objects

henkhooft committed:
Add gameobjects for each map. Several bug fixes

henkhooft committed:
Tilemap editor bugfixes

henkhooft committed:
Several bugfixes and clean up

henkhooft committed:
Add removal of tiles and basics for adjacency connections

henkhooft committed:
Alter the Direction enum used and add more stuff for the adjacency connectors

henkhooft committed:
- Make save file based on loaded scene
- Add UI for placing tiles in the editor
- Several bugfixes

henkhooft committed:
Introduce chunks again

henkhooft committed:
Several bugfixes related to the chunking and the editor

henkhooft committed:
Rename layer and add sub layer index methods

henkhooft committed:
Incorporate sub layers for walls

henkhooft committed:
Add overwrite option to tilemap editor

henkhooft committed:
Implement the simple adjacency connector

henkhooft committed:
Allow building and a few bug fixes related to tile loading

henkhooft committed:
Add reinitialization of tiles and add gameobjects for layers

henkhooft committed:
Fix boombox and light bulb networking

henkhooft committed:
Resolve a bug for which multiple fixtures on a tile couldnt be reintialized

henkhooft committed:
Solve several bugs related to multi tile object placement
Solve bug where icons don't load in the editor

henkhooft committed:
Fix several bugs related to tilemap initalization on the build version
Fix a bug in which multi tiled objects are not correctly loaded when on a chunk border

cosmiccoincidence commented:
This will also close #127 but you can only have 10 issues linked...

henkhooft committed:
Migrate all existing fixtures and improve icon loading in the editor

henkhooft committed:
Remove the existing tilemap code and port the majority of the code

henkhooft committed:
Small items changes

henkhooft committed:
Add restrictions to placing objects
Simplify the code a bit for setting tile objects

henkhooft committed:
Improve the multi adjacency connector

cosmiccoincidence commented:
Only worry about the simple, advanced, pipes, & offset pipes connectors.

The tiered ones are used for windows which I think the new method I'm working on will be much simpler and use the advanced one instead of needing tiered ones.

The wires were done not correctly and need to be reworked anyway to.

henkhooft committed:
Add simple and advanced connectors to the wooden table, metal wall and middle pipe layer

henkhooft committed:
Add the offset adjacency connector for pipes

joaoburatto commented:
**Let's ducking goooooooooooooo**

henkhooft committed:
Update missing files

henkhooft committed:
Merge remote-tracking branch 'SS3D_original/master' into tilemap-rework-2

# Conflicts:
#	Assets/Content/Furniture/Generic/Surfaces/Wood/Table - Wood.prefab
#	Assets/Content/Furniture/Machines/Medical/Cryotube - Blue.prefab
#	Assets/Content/Items/Functional/Tools/Generic/DeskLamp.cs
#	Assets/Content/Resources/Pipes/Pipe - Under L1.asset
#	Assets/Content/Resources/Pipes/Pipe - Under L1.asset.meta
#	Assets/Content/Resources/Pipes/Pipe - Under L3.asset
#	Assets/Content/Resources/Pipes/Pipe - Under L3.asset.meta
#	Assets/Content/Structures/Doors/Door.cs
#	Assets/Content/Structures/Pipes/Atmospherics/Pipe Layers/Pipe - Under L1.asset.meta
#	Assets/Content/Structures/Pipes/Atmospherics/Pipe Layers/Pipe - Under L1.prefab
#	Assets/Content/Structures/Pipes/Atmospherics/Pipe Layers/Pipe - Under L2.prefab
#	Assets/Content/Structures/Pipes/Atmospherics/Pipe Layers/Pipe - Under L3.asset.meta
#	Assets/Content/Structures/Pipes/Atmospherics/Pipe Layers/Pipe - Under L3.prefab
#	Assets/Content/Structures/Pipes/Atmospherics/PipeLayers/PipesLayer1.asset.meta
#	Assets/Content/Structures/Pipes/Atmospherics/PipeLayers/PipesLayer3.asset.meta
#	Assets/Content/Structures/Wall Mounts/Lights/Light Bulb Fixture.prefab
#	Assets/Content/Structures/Walls/Wall - Metal.prefab
#	Assets/Engine/Tile/Connections/AdvancedAdjacencyConnector.cs
#	Assets/Engine/Tile/Connections/OffsetPipesAdjacencyConnector.cs
#	Assets/Engine/Tile/Connections/PipesAdjacencyConnector.cs
#	Assets/Engine/Tile/Connections/SimpleAdjacencyConnector.cs
#	Assets/Engine/Tile/Connections/TieredAdjacencyConnector.cs
#	Assets/Engine/Tile/Connections/TieredAdvancedAdjacencyConnector.cs
#	Assets/Engine/Tile/Connections/WiresAdjacencyConnector.cs
#	Assets/Engine/Tile/State/FixtureStateMaintainer.cs
#	Assets/Engine/Tile/Tile.cs
#	Assets/Engine/Tile/TileRework/Connections/AdjacencyBitmap.cs

henkhooft committed:
Fix some issues created by the merge with master

henkhooft committed:
Improve the loading of icons in the editor

Ryan089 commented:
A couple of issues I identified playing with the TileMap Editor:

**Bug. Interaction with external programs while placing tiles in TileMap Editor.** While adding tiles in the TileMap Editor, it appears that you are unable to select other programs on your computer. This is particularly painful if you are trying to type review notes into a document! To replicate, open RE:SS3D Editor Tools > TileMap Editor, select a tile (such as "Catwalk") and then click "Add". Attempt to select a different (i.e. non-Unity) program on the taskbar. The program will either flash up briefly and disappear, or it will not appear at all. Focus will always be sent back to Unity, even if you minimize Unity.

**Bug. TileMap Editor does not apply tile restrictions when deleting PlacedTileObjects**. Although the TileMap editor does not allow you to place a PlacedTileObject in an invalid position, it allows an invalid position to be achieved through deleting the requirements. To replicate, add a single plenum to a blank tile on a tilemap. Change selected layer to turf, and add a wall on the plenum. Observe that you are not permitted to add a wall on any blank tile ("Cannot build here" message appears in console). Switch back to plenum layer, select "Delete" and delete the plenum. Note that the wall remains despite the plenum being deleted. Recommendation: all PlacedTileObjects on a tile are checked for validity once any PlacedTileObject on the tile is deleted, and they are also deleted if it is a prerequisite.

**Improvement. TileMap Editor should graphically indicate whether requirements are met when adding PlacedTileObjects**. Currently, a green square will indicate that you are trying to add a PlacedTileObject in a particular tile, and a red square is reserved for when you are trying to delete. However, it is counterintuitive that the square is green when you do not meet the requirements for placement. Recommendation: when adding PlacedTileObjects, use a green square to indicate that the object can be legally placed there, and an amber square to indicate that the object cannot be placed there.

Ryan089 commented:
**Bug. TileMap Editor "Allow overwrite" can result in incompatible PlacedTileObjects.** "Allow Overwrite" permits you to replace an object already in a layer. However, it does not check for conflicting objects on that tile, so you can create illegal arrangements. To replicate, go to Furniture Base layer and add a table onto a plenum. Next, select Turf layer and attempt to add a wall onto the same tile as the table (with "Allow overwrite" unchecked). Observe that the wall cannot be placed. Now, check the "Allow Overwrite" and attempt it again. Observe that the wall was placed in the same tile as the existing table, thus intersecting each other. Recommendation: all other PlacedTileObjects on a tile are checked for validity once any PlacedTileObject is added using the "Allow overwrite" functionality is enabled, and they are deleted if it they are no longer valid.

**Bug. TileMap Editor appears to behave indeterminately after a second TileManager has been added to the scene.** With only the initial TileManager in the scene, the TileMap Editor behaves predictably. However, once a second TileManager has been added, any one of a multitude of problems can manifest. These include the TileMap Editor window opening but being completely empty, the TileMap Editor opening correctly but being unresponsive to object placement, and the TileMap Editor duplicating all TileMaps from a different TileManager. Note that this erratic behaviour will continue even if the additional TileManagers have been deleted from the scene. I suspect that the root cause of the issue is that TileManager is intended to be a singleton, but the pattern has not been implemented correctly.

henkhooft committed:
Fix bug where the editor always sets focus and add an indicator if an object can be build to the editor

henkhooft committed:
Add optional connections to the multi adjacency connector

henkhooft committed:
Add networking for the optional adjacency connections

henkhooft committed:
Set more restrictions when deleting or replacing objects

henkhooft committed:
Solve initialization bug and check for multiple instances of the tilemanager

Ryan089 commented:
Keep up the great work! I've got another couple of minor issues for you to check out:

**Bug. When adding PlacedTileObjects with the TileMap Editor, chunk boundaries don't disappear**. Once you have selected "Add" in the TileMap Editor, moving your mouse over the scene will cause a yellow outline of any chunk the mouse passes over. However, the outline isn't removed once the mouse is over a different chunk. This issue doesn't appear to affect functionality in any way, it just looks a little strange.

**Bug. TileMap Editor now allows some incompatible objects to be placed together**. You can now add Walls and Tables on the same tile, even without "Allow Overwrite" checked, which shouldn't be possible. This error wasn't present in the last commit I reviewed.

henkhooft committed:
Fix tile adjacency networking and the crowbar

henkhooft committed:
Add more restrictions for walls and furniture

henkhooft committed:
Fix bug in which empty chunks are created but not removed

henkhooft committed:
Remove duplicate code in TileRestrictions

cosmiccoincidence commented:
Something to note with my connections design is that it will require the ability for multiple objects on a prefab to have different adjacency connectors.

The wall prefab would have pieces on it associated with the grille, plating, window divider, construction stage piece, etc.. but these pieces still need to be shaped according to adjacent connections as well. Most of these pieces would require the same connection script as the root object and could maybe just inherit from it, but the window divider piece would definitely need a unique connection script.

Is this possible?

henkhooft commented:
> Something to note with my connections design is that it will require the ability for multiple objects on a prefab to have different adjacency connectors.
> 
> The wall prefab would have pieces on it associated with the grille, plating, window divider, construction stage piece, etc.. but these pieces still need to be shaped according to adjacent connections as well. Most of these pieces would require the same connection script as the root object and could maybe just inherit from it, but the window divider piece would definitely need a unique connection script.
> 
> Is this possible?

In theory it would work well. The adjacency connector on the root object can call any connector in a child object

henkhooft committed:
Rework how adjacencies are edited and networked
Remove a few deleted scripts on prefabs

henkhooft committed:
Get cross connect adjacencies working for airlocks

henkhooft committed:
Improve wrench to include option for rotation

henkhooft committed:
Comment all files

cosmiccoincidence reviewed (changes requested):
- Not sure if you can do something about this or not, but the box to enter a name for the tilemap, is rather small, it would be more effective if it were wider.
![image](https://user-images.githubusercontent.com/41941008/126212530-89d9ce4d-fd7f-443f-9189-d2e176d45c2f.png)

- The custom tilemap name only registers if you hit "New" while you are still selected on the name field. It will also still say "Empty map (1)" as the active tilemap.
https://user-images.githubusercontent.com/41941008/126213006-cae10bc6-7b23-4516-b5b9-5a7ae61d31d9.mp4

- Selecting "Is Main Map: []" prior to hitting "New" will not register.

EDIT: I think I realize now that you need to hit "apply".

cosmiccoincidence reviewed (changes requested):
- Am I getting an error? I cannot see the images of the prefabs in the tilemap editor.
![image](https://user-images.githubusercontent.com/41941008/126214650-2a1f3887-3ffc-4c29-be1f-4a363f403ecf.png)

- Hold+Shift no longer works to lay a large area of tiles at once.

- Walls/Windows lost their advanced adjacency scripts.

- Though I wont deem it necessary in this PR, we should improve the 'selected layer' drop-down menu. Currently, it lists the bottom-most layer at the top and works its way up through the layers while going down the menu, this should be flipped. We should also maybe color-code or find another way to help users visually identify the difference between layers residing in the base and layers residing above it (these should be separated into floor + fixtures and wall + fixtures).Ex. 
![image](https://user-images.githubusercontent.com/41941008/126216318-57e2d43f-1eeb-4f4c-b352-eb493812716f.png)

DoomsdayOrange commented:
You should be able to replace floors and walls without deleting the original first. Also there should be a key you hold down to enter delete mode instead of clicking a button.

cosmiccoincidence reviewed (changes requested):
- Since you are using yellow to represent chunk boundaries, I recommend changing the 'invalid location' color from yellow to red when placing tiles.
![image](https://user-images.githubusercontent.com/41941008/126217373-3a1514b6-b160-4b00-afa9-94267b39a5c5.png)

- You can put tiles at the same position on different tilemaps and can overlap furniture and other objects this way. I haven't tried playtesting the game to see what would happen there but presumably, we should have some type of tilemap collision or error.
![image](https://user-images.githubusercontent.com/41941008/126218913-16ac9caa-923f-45d5-bb20-9d01b094ab7f.png)

I haven't finished reviewing this yet, but these are just some things I noticed on my first quick look.

cosmiccoincidence commented:
> You should be able to replace floors and walls without deleting the original first.

Enable the "Allow Overwrite" feature.

DoomsdayOrange commented:
> > You should be able to replace floors and walls without deleting the original first.
> 
> Enable the "Allow Overwrite" feature. @DoomsdayOrange 

Ah whoops

Ryan089 commented:
> * Since you are using yellow to represent chunk boundaries, I recommend changing the 'invalid location' color from yellow to red when placing tiles.

Probably not red, as that is used for removing tiles. Orange perhaps?

cosmiccoincidence commented:
> > * Since you are using yellow to represent chunk boundaries, I recommend changing the 'invalid location' color from yellow to red when placing tiles.
> 
> Probably not red, as that is used for removing tiles. Orange perhaps?

Good point. Alternatively we could leave it yellow and change the chunk border to another color (like blue).

Ryan089 commented:
**Bug. Many prefabs still have the deleted FixtureStateMaintainer scripts.** These now appear as Missing (Mono Script) on the inspector, and flag multiple warnings on scene load. Prefabs believed to have these scripts are detailed below. Recommend removing this component from all applicable prefabs.

![Prefabs with FixtureStateMaintainer class](https://user-images.githubusercontent.com/64360820/126327086-8404396e-fbaf-4a5c-ac47-da44f7ffbf15.JPG)

cosmiccoincidence commented:
Did we ever come to conclusion on the resources folders thing?

cosmiccoincidence commented:
Can't place a 'Light Bulb Fixture' on the map via editor. The reason for this is due to an extra /Wall Mounts/ folder. The related TileObjectSO is connected to the (I'm assuming) old light bulb fixture prefab in this extra folder. Removing these extra folders and files and connecting the TileObjectSO to the correct prefab will solve this issue.

![image](https://user-images.githubusercontent.com/41941008/126675179-940391c2-b3be-44aa-96fd-7f6272f1566d.png)

henkhooft committed:
Remove FixtureStateMaintainer

henkhooft committed:
Implement the tile drag handler and fix an issue in which objects on multiple tile objects can collide

henkhooft committed:
Fix bug in which the adjacency connector wouldn't update if a wall was destroyed via a missing plenum
Removed the faulty light bulb prefab

henkhooft committed:
Add adjacency connectors to walls and cable

cosmiccoincidence commented:
Known Issue: You can see inside of walls thru windows. This is an issue with the adjacency connectors for windows which can be handled in a later task.

![image](https://user-images.githubusercontent.com/41941008/126735981-2e8678f1-45e8-43f0-a112-a7c2f14f7f7b.png)

cosmiccoincidence commented:
Known Issue: Wires do not connect as they have a complex design and need a new script made, outside of this task.

![image](https://user-images.githubusercontent.com/41941008/126736529-e40bd2ed-8f0e-41f6-853b-df980d8a9697.png)

cosmiccoincidence reviewed (changes requested):
- Disposal Pipe prefab is missing its simple adjacency connector.

- Atmos pipes are missing their mesh filters and assigned adjacency meshes.

- L2 Atmos pipe has the Pipe Object script but the others don't?

- I'd add the word "Mount" after the "HighWall" and "LowWall" to avoid any confusion.
![image](https://user-images.githubusercontent.com/41941008/126736634-4f132f2e-be84-48fc-8aec-49339a4a1370.png)

- Walls seem to want to connect to doors from all directions. Preferably they should only want to try to connect to them via their 'sides' otherwise you get odd things like this.
![image](https://user-images.githubusercontent.com/41941008/126737486-de611dfd-1694-4da6-8120-556caaaf331c.png)

- Girders won't place on the map, you need to add 'wall' as their type in both their prefab and their TileObjectSO.

- Speaking of types. I actually changed these back in a previous PR I believe so they would all be capitalized, this way we could have ones like "WallMount", "FloorMount", "DisposalPipe" that are more readable than if they were completely lowercase. Unless there is an issue, I would like to see them capitalized again.

- There is an extra copy of the metal wall prefab and tile object in the /Walls/ folder. It seems the one that is correctly named (In PascalCase) is the one that is not correctly set up and has the incorrect tile object.

- Speaking of your TileObjectSOs, those should also be properly named with PascalCase. Rename these according to their related prefabs.

My Unity died on my while reviewing so wrap up these things and I'll come back for another later.

henkhooft committed:
Add disposal pipe adjacency and set wire generic type

henkhooft committed:
Fix wire adjacencies and clean a bunch of assets
Add "mount" to the low and high wall enum

henkhooft committed:
Update table adjacencies

henkhooft committed:
Resolve multiple bugs in which doors connect to walls when not intended

henkhooft committed:
Create MiniStation and remove the old maps
Remove duplicate metal wall

cosmiccoincidence reviewed (changes requested):
- Cryotube.prefab seems like another rouge duplicate prefab.

- Plenum.prefab missing simple adjacency connector.

- I'm not really sure how to edit a pre-existing map, when I go to the map editor it doesn't have the ministation map as active and when I hit load it actually deleted it...

- Speaking of the map, it would be nice if you add a 2nd tilemap to the new ministation to demonstrate 2 tilemaps in the same map. (also add some items)

- Building tables on either the client or host only builds them on the host.
![image](https://user-images.githubusercontent.com/41941008/126852899-698cfb49-4eda-47c7-ab33-9f057b98aac0.png)

- Not sure if something you can do about or not but doing a build and running 2 clients was pretty laggy, door one client would open several seconds after it was opened on the original client. Seems dropping items on the joining client would rubberband the item for me a bit too.

- Wouldn't hurt to merge master in, not sure when you did last.

henkhooft committed:
Expand MiniStation

henkhooft committed:
Remove cryo tube duplicate
Add adjacency connector to plenum
Remove stale atmos object
Make wood table spawnable

henkhooft committed:
Merge remote-tracking branch 'SS3D_original/master' into tilemap-rework-2

henkhooft committed:
Add network transform to hardhats

henkhooft commented:
> 
> 
>     * Cryotube.prefab seems like another rouge duplicate prefab.
> 
>     * Plenum.prefab missing simple adjacency connector.
> 
>     * I'm not really sure how to edit a pre-existing map, when I go to the map editor it doesn't have the ministation map as active and when I hit load it actually deleted it...
> 
>     * Speaking of the map, it would be nice if you add a 2nd tilemap to the new ministation to demonstrate 2 tilemaps in the same map. (also add some items)
> 
>     * Building tables on either the client or host only builds them on the host.
>       ![image](https://user-images.githubusercontent.com/41941008/126852899-698cfb49-4eda-47c7-ab33-9f057b98aac0.png)
> 
>     * Not sure if something you can do about or not but doing a build and running 2 clients was pretty laggy, door one client would open several seconds after it was opened on the original client. Seems dropping items on the joining client would rubberband the item for me a bit too.
> 
>     * Wouldn't hurt to merge master in, not sure when you did last.



> 
> 
>     * Cryotube.prefab seems like another rouge duplicate prefab.
> 
>     * Plenum.prefab missing simple adjacency connector.
> 
>     * I'm not really sure how to edit a pre-existing map, when I go to the map editor it doesn't have the ministation map as active and when I hit load it actually deleted it...
> 
>     * Speaking of the map, it would be nice if you add a 2nd tilemap to the new ministation to demonstrate 2 tilemaps in the same map. (also add some items)
> 
>     * Building tables on either the client or host only builds them on the host.
>       ![image](https://user-images.githubusercontent.com/41941008/126852899-698cfb49-4eda-47c7-ab33-9f057b98aac0.png)
> 
>     * Not sure if something you can do about or not but doing a build and running 2 clients was pretty laggy, door one client would open several seconds after it was opened on the original client. Seems dropping items on the joining client would rubberband the item for me a bit too.
> 
>     * Wouldn't hurt to merge master in, not sure when you did last.
I've made the changes as requested. I'm not sure why you cannot edit MiniStation anymore. Does the save file ministation.txt exist under the StreamingAssets/Saves folder?

The reason why the tables didn't spawn was due to them not being in the spawnable prefab list. I've already made a mention of it in the limitations, so to the make it future proof we need to link it to addressable assets.

The rubberbanding behaviour I've noticed on previous builds as well. I believed it was caused due to the framerate being limited if an instance is running in the background.

Ryan089 commented:
**Overlays.** It appears that there is now only a single "Overlay" layer, rather than the three overlay layers previously. This prevents us from creating (for example) the blue tiled border within the medbay room on MainStation. Is it possible to make the number of overlays on a tile unlimited? If not, I think four overlay layers would be convenient.

Ryan089 commented:
**Visual indicator when using "Allow Overwrite"**. When using Allow Overwrite in the Tilemap Editor, the coloured border is yellow when hovering over a tile that already has a PlacedTileObject in that layer, indicating that the proposed object cannot be placed there. However, provided all other availability criteria are met, the border should be green because the existing PlacedTileObject can be overwritten.

henkhooft committed:
Solve an edge case bug in which the chunks dictionary was modified during map deletion

henkhooft committed:
Fix a bug in which the rotion of the ghost object isn't updated when switching objects

henkhooft committed:
Add multiple layers for overlays

henkhooft committed:
Fix issue where tile override is not showing in the editor if an object can build

henkhooft commented:
Another round of bugs solved 👍. Though I noticed that there is some Z-fighting going on between overlays and floor tiles.

Ryan089 commented:
**TileRestrictions.CanBuild treatment of plenum layer.** Currently, catwalks and lattices are treated identically to plenums, so you can build walls, tables etc on them. This is contrary to the comments on Cosmic's Tilemap Layers:

> Lattice, Catwalk, & Plenum all have different things that can be built on them... The way I see it is. A lattice is made first, and nothing can be built ON it. You must first add metal and build it into a catwalk or plenum. A Catwalk can have wires, above floor pipes, and machines built on it, but nothing else.

 I'm unsure if this is a actually a bug, or simply a design decision.

Ryan089 commented:
> Though I noticed that there is some Z-fighting going on between overlays and floor tiles.

We'd resolved the Z-fighting on the previous Tilemap system by adding a small increment to the y position of the overlay. This increment still exists on the overlay prefabs, so I think the issue could be solved by amending the calculation of the placedObjectWorldPosition in Tilemap.SetTileObject:

`Vector3 placedObjectWorldPosition = chunk.GetWorldPosition(placedObjectOrigin.x, placedObjectOrigin.y) + tileObjectSO.prefab.transform.position + new Vector3(rotationOffset.x, 0, rotationOffset.y) * chunk.GetTileSize();`

Of course, we'd have to be vigilant in ensuring all prefabs are set to the origin except where an offset is explicitly required.

henkhooft committed:
Take prefab position into account when placing objects to avoid Z-fighting for overlays

henkhooft committed:
Add more restrictions to plenums

henkhooft committed:
Solve issue in which icons weren't loading in the editor

henkhooft committed:
Solve bug in which overlays with a rotation applied weren't correctly loaded

henkhooft committed:
Update MiniStation to remove a few illegally placed objects

cosmiccoincidence commented:
I went through all the issues that this PR is supposed to close and tested them each, it seems they all are fixed now.

Regarding the tilemap design itself #682 it seems you have resolved the edge case issues, and it seems the construction issues are now possible but yet to be implemented... but what about the moving tilemaps is that possible at this current state? and is there a way for me to test it?

![image](https://user-images.githubusercontent.com/41941008/128278393-c9e1376a-dddf-4a98-bdf7-c806bbe77139.png)

It seems you also handled the 'other additions' listed in #682 but the last thing I want to ask is, would it be hard to add marching squares in the future? Once we have the assets, it should be possible to add the marching squares without a bunch of reworking the tilemap system.

cosmiccoincidence reviewed (commented):
- We have 2 wood table prefabs, TableWood.prefab has the correct name but the other one has the correct adjacency connectors.

- You moved TankStorageUnit.prefab from  /Content/Furniture/Machines/Atmospherics/  to  /Content/Furniture/Machines/Atmospherics/Canisters/  please move this back.

- Looks like something wacky happened with a merge at some point. The TableLamp script got upgraded and the lamp files moved (now located at  /Content/Items/Functional/Generic/TableLamp.cs) but it looks like the namespace in the script didn't get updated when it got moved. It also looks like there is a copy of the old script (/Content/Items/Functional/Tools/Generic/DeskLamp.cs).

cosmiccoincidence reviewed (changes requested):
- Remove extra .asset files for botany plants. Plant prefabs got consolidated down from individual prefabs for each plant stage (6 total) to just 1 prefab per plant type. We no longer need 6 TileObject asset files for each plant, the extras should be removed and the remainder should be correctly linked to their prefabs.

- Similarly, I noticed these TileObject asset files are not properly named after a recent asset renaming PR, please click on these files and view the linked prefab and rename the asset file to match the prefab file.

- 'Wall - Metal.prefab' should be named 'MetalWall.prefab', this is likely a merge error.

cosmiccoincidence commented on Assets/Engine/Tile/TileRework/Connections/AdjacencyTypes/OffsetAdjacency.cs:
```diff
@@ -0,0 +1,136 @@
+        public enum OffsetOrientation
+        {
+            o,
+            cNorth,
```
the 'c' meshes in the adjacency connectors got renamed to 'u' in a previous PR, this seems to be another merge error. It seems that the AdvancedAdjacency.cs was merged correctly and uses 'u', this file didn't get that memo.

This should be 'uNorth'

cosmiccoincidence commented on Assets/Engine/Tile/TileRework/Connections/AdjacencyTypes/OffsetAdjacency.cs:
```diff
@@ -0,0 +1,136 @@
+        {
+            o,
+            cNorth,
+            cSouth,
```
uSouth

cosmiccoincidence commented on Assets/Engine/Tile/TileRework/Connections/AdjacencyTypes/OffsetAdjacency.cs:
```diff
@@ -0,0 +1,136 @@
+        [Tooltip("A mesh where no edges are connected")]
+        public Mesh o;
+        [Tooltip("A mesh where the north edge is connected, can be rotated to the east")]
+        public Mesh cNorth;
```
uNorth

cosmiccoincidence commented on Assets/Engine/Tile/TileRework/Connections/AdjacencyTypes/OffsetAdjacency.cs:
```diff
@@ -0,0 +1,136 @@
+        [Tooltip("A mesh where the north edge is connected, can be rotated to the east")]
+        public Mesh cNorth;
+        [Tooltip("A mesh where the south edge is connected, can be rotated to the west")]
+        public Mesh cSouth;
```
uSouth

cosmiccoincidence commented on Assets/Engine/Tile/TileRework/Connections/AdjacencyTypes/OffsetAdjacency.cs:
```diff
@@ -0,0 +1,136 @@
+                mesh = o;
+                orientation = OffsetOrientation.o;
+            }
+            else if (cardinalInfo.IsC())
```
IsU

cosmiccoincidence commented on Assets/Engine/Tile/TileRework/Connections/AdjacencyTypes/OffsetAdjacency.cs:
```diff
@@ -0,0 +1,136 @@
+            {
+                if (cardinalInfo.north > 0 || cardinalInfo.east > 0)
+                {
+                    mesh = cNorth;
```
uNorth (same with line below)

cosmiccoincidence commented on Assets/Engine/Tile/TileRework/Connections/AdjacencyTypes/OffsetAdjacency.cs:
```diff
@@ -0,0 +1,136 @@
+                }
+                else
+                {
+                    mesh = cSouth;
```
uSouth (same with line below)

cosmiccoincidence commented on Assets/Engine/Tile/TileRework/Connections/AdjacencyTypes/OffsetAdjacency.cs:
```diff
@@ -0,0 +1,136 @@
+            return new MeshDirectionInfo { mesh = mesh, rotation = rotation };
+        }
+    }
+}
```
Missing blank line at end of file

cosmiccoincidence commented on Assets/Engine/Tile/TileRework/Connections/AdjacencyTypes/SimpleAdjacency.cs:
```diff
@@ -0,0 +1,64 @@
+        [Tooltip("A mesh where no edges are connected")]
+        public Mesh o;
+        [Tooltip("A mesh where the north edge is connected")]
+        public Mesh c;
```
u

cosmiccoincidence commented on Assets/Engine/Tile/TileRework/Connections/AdjacencyTypes/SimpleAdjacency.cs:
```diff
@@ -0,0 +1,64 @@
+                mesh = o;
+            else if (cardinalInfo.IsC())
+            {
+                mesh = c;
```
u

cosmiccoincidence commented on Assets/StreamingAssets/Saves.meta:
```diff
@@ -0,0 +1,8 @@
+fileFormatVersion: 2
```
Any reason for this folder its files to be where it is? We try to fit everything in one of the 4 classes (art, content, editor, engine), these seem fit for maybe the content folder instead of the asset folder directly?

henkhooft replied:
Unfortunately before the tilemap save file wasn't being included in builds. This folder is always being copied by Unity into a build. https://docs.unity3d.com/Manual/StreamingAssets.html

cosmiccoincidence replied:
So the StreamingAssets folder is required to be in this location and cannot be moved?

henkhooft replied:
Yeah it's generated by Unity automatically. Though I'm not sure if another method exists to include the tilemap save file with builds.

henkhooft commented:
> 
> 
> I went through all the issues that this PR is supposed to close and tested them each, it seems they all are fixed now.
> 
> Regarding the tilemap design itself #682 it seems you have resolved the edge case issues, and it seems the construction issues are now possible but yet to be implemented... but what about the moving tilemaps is that possible at this current state? and is there a way for me to test it?
> 
> ![image](https://user-images.githubusercontent.com/41941008/128278393-c9e1376a-dddf-4a98-bdf7-c806bbe77139.png)
> 
> It seems you also handled the 'other additions' listed in #682 but the last thing I want to ask is, would it be hard to add marching squares in the future? Once we have the assets, it should be possible to add the marching squares without a bunch of reworking the tilemap system.

Moving tilemaps are in theory possible as all tiles are a child of the map object. So moving the object will also move the tiles. There is one issue with it right now and that is that placing/removing tiles is world position based. I would have to make some adjustments to take the origin position and rotation as a basis for placing and removing tiles.

On the marching squares, the challenges I see are in it having a floor and wall at the same time and how to spawn them when required. Normally placing walls and floors is not possible at the same time, so a potential solution/hack for it would be to include a floor into the marching squares. Alternatively a sub layer needs to be created for walls and floors, with a bunch of restriction rules programmed in to not allow walls being placed on floors for the majority of cases.

Having angled tiles themselves can be easily achieved with an adjacency connector and having a reduced atmos volume as well.

cosmiccoincidence commented:
> On the marching squares, the challenges I see are in it having a floor and wall at the same time and how to spawn them when required. Normally placing walls and floors is not possible at the same time, so a potential solution/hack for it would be to include a floor into the marching squares. Alternatively a sub layer needs to be created for walls and floors, with a bunch of restriction rules programmed in to not allow walls being placed on floors for the majority of cases.

Alternatively, with the fact that a half floor would be too small to support machines and such, I think we can add the angled floor model directly on the angled wall model. This way we don't need to solve the issue of having both a floor and a wall on the same tile, it would be effectively treated as just a (special) wall.

henkhooft committed:
Remove duplicate wooden table and change location of tank storage

henkhooft committed:
Remove unnecessary plant stages

henkhooft committed:
Remove old C connections for adjacencies and rename the metal wall

henkhooft committed:
Update plant assets

henkhooft committed:
Fix table lamp

cosmiccoincidence commented:
- Make sure ALL the TileObjectSO files are correctly named AND the name string is correct. Both of those should be the same as the prefab names they reference.

![image](https://user-images.githubusercontent.com/41941008/128935812-d290a495-2db4-4270-ab27-8a5045f5c739.png)

cosmiccoincidence commented:
Another thing I would like added for a while now but keep forgetting to write it down out is:

Add 'Specific Type' as another option that goes with the 'Generic Type' thing.
![Unity_GaMuDuYDJo](https://user-images.githubusercontent.com/41941008/128968137-013ea8be-5f8c-4fde-a157-f2a779edb88f.png)

I foresee us being able to use the specific type to accommodate some specific connection cases. Examples:

- We don't want different types of tables to connect to each other. I believe we currently prevent this by having different generic types for each table type, but preferably they would all have "table" as their generic type and we can give them unique specific types to prevent connections.

- Currently, we only have 1 type of wall (metal) and thus only 1 type of door frame (metal). Once we have more types of walls (alien, gold, wood, etc.), we will also need door frames of the same type and a method to determine what type of door frame is required to connect to the adjacent walls. Specific types could be used to determine what door frame is needed. (Note: unlike tables, walls CAN connect to different types of walls.)

henkhooft committed:
Update TileObjectSOs to new naming standard

henkhooft committed:
Add specific type to adjacency connector

henkhooft committed:
Add tile migrator

henkhooft commented:
> 
> 
> Another thing I would like added for a while now but keep forgetting to write it out is:
> 
> Add 'Specific Type' as another option that goes with the 'Generic Type' thing.
> ![Unity_GaMuDuYDJo](https://user-images.githubusercontent.com/41941008/128968137-013ea8be-5f8c-4fde-a157-f2a779edb88f.png)
> 
> I foresee us being able to use the specific type to accommodate some specific connection cases. Examples:
> 
>     * We don't want different types of tables to connect to each other. I believe we currently prevent this by having different generic types for each table type, but preferably they would all have "table" as their generic type and we can give them unique specific types to prevent connections.
> 
>     * Currently, we only have 1 type of wall (metal) and thus only 1 type of door frame (metal). Once we have more types of walls (alien, gold, wood, etc.), we will also need door frames of the same type and a method to determine what type of door frame is required based on the connected wall. Specific types could be used to determine what door frame is needed. (Note: unlike tables, walls CAN connect to different types of walls.)

As suggested, I have added an option to specify a "specific" identifier to a TileObjectSO. Interesting enough I discovered that the adjacency connector can be a bit simplified, so now you no longer have to the specify what types it connects to in the adjacency connector.

![afbeelding](https://user-images.githubusercontent.com/5231626/129781654-9918a940-bf81-4882-8a17-22f201dc8dd9.png)

henkhooft committed:
Remake MiniStation

henkhooft commented:
There is a limitation to make items part of the tilemap. Adding them under the corresponding gameobject will unfortunately delete them during loading of map as items are not part of the save file. 

They can be integrated in the future, but will require some custom work such as applying less strict restrictions on placement positions and rotations. I would suggest to add that as a future feature.

henkhooft committed:
Fix missing meshes on pipes and plenum

henkhooft committed:
Fix bug in which plenums did not connect under walls

cosmiccoincidence committed:
Merge branch 'master' into pr/736

cosmiccoincidence committed:
Removal of extra meta file

cosmiccoincidence committed:
Capitalize Fridge

Unity did this on its own

cosmiccoincidence committed:
Fix Cables and Disposal missing meshes

cosmiccoincidence reviewed (approved):
Well, I don't know what else to look for or request with this...
You've really done god's work with this.
I'm going to leave #682 open for now until you finish the new tilemap design document.
